### ISSUE TITLE
Spelunk and parser breaks with this code

### ISSUE THREAD

**Issue by JPaulsen** [dart-lang/linter#478]
I am trying to reproduce this bug https://github.com/dart-lang/linter/issues/473#issue-212261603, and I noticed that the visitor breaks when trying to do it, With the debugger I noticed that it is not parsing well the AST in this part, and I created the minimal code that breaks the spelunk, and this is it:

```
Iterable stateList(Finder finder) {
  return finder.evaluate().map((Element element) => _stateOf(element, finder));
}

class State {
}

class Finder {
  Iterable<Element> evaluate() => [];
}

class StatefulWidget {
}

class Element {
}

T _stateOf<T>(Element element, Finder finder) {
  return null;
}
```

The error message is this:

```
Unhandled exception:
Exception: [test/rules/unnecessary_lambdas_falsepositive.dart(247..247): Functions must have an explicit list of parameters, test/rules/unnecessary_lambdas_falsepositive.dart(247..247): A function body must be provided, test/rules/unnecessary_lambdas_falsepositive.dart(247..247): Expected a method, getter, setter or operator declaration, test/rules/unnecessary_lambdas_falsepositive.dart(247..247): Unexpected token '<', test/rules/unnecessary_lambdas_falsepositive.dart(249..249): Expected a method, getter, setter or operator declaration, test/rules/unnecessary_lambdas_falsepositive.dart(249..249): Expected a method, getter, setter or operator declaration, test/rules/unnecessary_lambdas_falsepositive.dart(249..249): Unexpected token '>', test/rules/unnecessary_lambdas_falsepositive.dart(268..273): Expected to find ';', test/rules/unnecessary_lambdas_falsepositive.dart(281..281): Expected a method, getter, setter or operator declaration, test/rules/unnecessary_lambdas_falsepositive.dart(281..281): Expected a method, getter, setter or operator declaration, test/rules/unnecessary_lambdas_falsepositive.dart(281..281): Unexpected token ')']
#0      _ErrorListener.throwIfErrors (package:analyzer/src/lint/util.dart:117:7)
#1      Spelunker.spelunk (package:analyzer/src/lint/util.dart:100:19)
#2      main.<anonymous closure> (file:///usr/local/google/home/casanueva/Development/linter/tool/spelunk.dart:13:54)
#3      Object&ListMixin.forEach (dart:collection/list.dart:63)
#4      main (file:///usr/local/google/home/casanueva/Development/linter/tool/spelunk.dart:13:16)
#5      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:259)
#6      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
```

**Comment by bwilkerson** [comment]
The linter package is currently behind bleeding edge. In order to use a version of analyzer that has support for generic methods you will need to either (a) use the linter embedded in a tool such as the command-line analyzer or (b) add a dependency override to the linter's pubspec.yaml file that points both 'analyzer' and 'front_end' to the bleeding edge versions in a checkout of the 'sdk' repository.

**Comment by pq** [comment]
Thanks Brian. 👍 

Feel free to ping me @JPaulsen if you need a hand getting this setup.
